File: tools/forge/bakeoff/src/sdfpy/B5_candelabra.py
"""B5 candelabra — the grammar item.

The tree is an ordinary recursive Python function that APPENDS primitives to a
list; nothing is instanced, transformed or joined.  At the end one
`union(*limbs, k=BLEND)` turns ~50 tapered capsules into a single organism,
and every junction is a smooth minimum, so "must not show open seams" is not
something that has to be arranged — a seam is not expressible here.

Reuse is by calling the function again with different arguments; the taper is
literally `radius * TAPER` on the recursive call.
"""

import os
import sys

import numpy as np
from sdf import Z, rounded_cylinder, slab, sphere, union

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import sdfkit as kit
from sdfkit import tapered_capsule

BASE_R, BASE_H = 0.70, 0.16
TRUNK_R, TRUNK_H = 0.22, 1.20
TAPER = 0.75  # radius factor per generation
ARMS, FORKS = 3, 2
SPREAD = np.radians(41)  # gen-1 lift-off angle from vertical (spec: 35-45)
FORK = np.radians(38)  # gen-2 half-angle, swung tangentially
UPNESS = 0.38  # how hard an arm curls back toward vertical
ARM_LEN, FORK_RATIO = 1.07, 0.70
BLEND = 0.075  # smooth-union radius where a branch meets its parent
SEG_BLEND = 0.02  # inside an arm the segments already meet, so barely any

PAN_R, PAN_H = 0.28, 0.07
CUP_R, CUP_BORE, CUP_H = 0.135, 0.065, 0.21  # 0.07 wall, at the visibility floor

limbs = []


def _unit(v):
    return np.asarray(v, float) / np.linalg.norm(v)


def _spin(v, axis, angle):
    """Rodrigues rotation of `v` about `axis` by `angle`."""
    v, k = np.asarray(v, float), _unit(axis)
    return v * np.cos(angle) + np.cross(k, v) * np.sin(angle) + k * (k @ v) * (1 - np.cos(angle))


def _sconce(tip):
    """Drip-pan plus candle cup, dished and bored by two more primitives."""
    z = tip[2]
    pan = rounded_cylinder(PAN_R, 0.025, PAN_H).translate((tip[0], tip[1], z + PAN_H / 2))
    dish = sphere(1.2, (tip[0], tip[1], z + PAN_H + 1.165))  # shallow: big radius
    cup = rounded_cylinder(CUP_R, 0.02, CUP_H).translate((tip[0], tip[1], z + CUP_H / 2))
    bore = rounded_cylinder(CUP_BORE, 0.015, CUP_H).translate((tip[0], tip[1], z + CUP_H / 2 + 0.09))
    return ((pan - dish) | cup) - bore


def grow(origin, direction, length, radius, depth, steps=6):
    """Sweep one arm as a chain of round cones, then fork or finish.

    Returns the SDF for this whole subtree, so the recursion composes fields
    rather than mutating a global list.
    """
    p, d = np.asarray(origin, float), _unit(direction)
    segments = []
    for i in range(steps):
        lean = _unit(d + (Z - d) * (UPNESS * (i + 1) / steps))  # curl toward vertical
        nxt = p + lean * (length / steps)
        r0 = radius * (1 - (1 - TAPER) * i / steps)
        r1 = radius * (1 - (1 - TAPER) * (i + 1) / steps)
        segments.append(tapered_capsule(tuple(p), tuple(nxt), r0, r1))
        p, d = nxt, lean
    # consecutive segments already share an endpoint AND a radius, so joining
    # them with the junction blend would only bead the arm.  Barely blend.
    arm = union(*segments, k=SEG_BLEND)

    if depth == 0:
        return union(arm, _sconce(p), k=SEG_BLEND)
    radial = _unit([d[0], d[1], 0]) if np.hypot(d[0], d[1]) > 1e-9 else np.array([1.0, 0, 0])
    children = [  # swing the children tangentially so the 6 tips ring evenly
        grow(p, _spin(d, radial, FORK * (2 * i / (FORKS - 1) - 1)),
             length * FORK_RATIO, radius * TAPER, depth - 1)
        for i in range(FORKS)
    ]
    return union(arm, *children, k=BLEND)


# --- base and trunk, then three arms ----------------------------------------
limbs.append(rounded_cylinder(BASE_R, 0.05, BASE_H).translate(Z * (BASE_H / 2)))
limbs.append(tapered_capsule((0, 0, BASE_H * 0.4), (0, 0, BASE_H + TRUNK_H), TRUNK_R * 1.25, TRUNK_R))

for i in range(ARMS):
    az = 2 * np.pi * i / ARMS
    lift = _spin(Z, [-np.sin(az), np.cos(az), 0], SPREAD)  # tilt out by SPREAD
    limbs.append(grow((0, 0, BASE_H + TRUNK_H), lift, ARM_LEN, TRUNK_R * TAPER, depth=1))

# the trunk's start cap is a sphere and pokes below the base; trim it flat so
# the model closes inside the sample box rather than against its wall
candelabra = union(*limbs, k=BLEND) & slab(z0=0.0)


if __name__ == "__main__":
    out = os.path.join(
        os.path.dirname(os.path.abspath(__file__)), "../../out/sdfpy/B5_candelabra.glb"
    )
    kit.report(
        "B5",
        kit.bake(
            candelabra,
            os.path.abspath(out),
            bounds=((-1.35, -1.35, -0.033), (1.35, 1.35, 3.42)),
            step=0.015,
            budget=20000,
            normals="field",
            angle=30.0,
        ),
    )

File: tools/forge/bakeoff/src/manifold/B5_candelabra.mjs
// Copyright 2026 The Dice Table Authors.
//
// B5 candelabra. Round base r 0.7, trunk r 0.22 rising 1.2, splitting into 3
// arms, each splitting into 2, a drip-pan and candle cup at all 6 tips, radii
// tapering x0.75 per generation, total height ~3.2.
//
// The grammar item, so the shape is written as a grammar: `grow()` calls
// itself and the ONLY thing it does is push tapered segments onto a list.
// Geometry happens once, afterwards. That split is what makes the recursion
// readable: no Manifold objects are created inside the recursion, so there is
// nothing to keep manifold at every step and nothing to leak.
//
// Junctions: the segment list becomes ONE sdf blended with a polynomial
// smooth-min, so every fork is a fillet by construction -- there are no seams
// to hide because there is no boolean in the branching at all. Drip pans and
// candle cups are the deliberate exception: they are turned metal parts, so
// they are exact CSG (dished cylinder, bored cup) unioned onto the tips, where
// a crisp rim is what the object should actually have.
import {Manifold} from 'manifold-3d/manifoldCAD';
import {describe, save, segDist, smin, stopwatch} from './_util.mjs';

const BASE_R = 0.7, BASE_H = 0.16;
const TRUNK_R = 0.22, TRUNK_TOP = 1.2;
const TAPER = 0.75;                // radius factor per generation
const ARMS = 3, FORKS = 2;
const SPREAD1 = 42, SPREAD2 = 46;  // degrees off the parent direction
const PAN_R = 0.28, PAN_H = 0.07;
const CUP_R = 0.115, CUP_H = 0.17, BORE_R = 0.072;
// Two density dials, and they are not equivalent. EDGE is the levelSet grid;
// SIMPLIFY_TOL is Manifold's own edge-collapse decimator, which is the better
// buy on a marching-tetrahedra mesh because it eats the sliver triangles the
// grid leaves behind rather than coarsening the shape. Measured on this model:
//   EDGE 0.068 raw                    -> 19.5k tris (coarse branches)
//   EDGE 0.048 + simplify(0.0035)     -> 17.3k tris from a 34.3k grid, and it
//                                        also collapses the sub-tolerance
//                                        pinches that put the raw 0.048 mesh
//                                        at genus 11
const EDGE = Number(process.argv[3] ?? 0.048);
const SIMPLIFY_TOL = Number(process.argv[4] ?? 0.0035);

// ---------------------------------------------------------------- grammar --
const segments = [];  // {a, b, r1, r2} tapered capsules, blended later
const tips = [];      // where a pan+cup assembly gets planted

const dot = (a, b) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
const add = (v, w, s = 1) => [v[0] + w[0] * s, v[1] + w[1] * s, v[2] + w[2] * s];
const norm = (v) => {
  const l = Math.hypot(...v) || 1;
  return [v[0] / l, v[1] / l, v[2] / l];
};

/** Tilt direction d by `deg` toward horizontal bearing `phi`. */
function tilt(d, deg, phi) {
  const rad = (deg * Math.PI) / 180;
  const u = [Math.cos(phi), Math.sin(phi), 0];
  const k = dot(u, d);
  const side = norm([u[0] - d[0] * k, u[1] - d[1] * k, u[2] - d[2] * k]);
  return norm(add(add([0, 0, 0], d, Math.cos(rad)), side, Math.sin(rad)));
}

/**
 * Grow one branch as a chain of short segments that curve back toward
 * vertical, then fork. Appends to `segments` / `tips`; returns nothing.
 * gen 1 = the three arms, gen 2 = the six forks that end in tips.
 */
function grow(origin, dir, length, radius, gen, bearing) {
  const STEPS = 5;
  let p = origin, d = dir;
  for (let i = 0; i < STEPS; i++) {
    d = norm([d[0], d[1], d[2] + 0.09]);  // lift a little every step
    const q = add(p, d, length / STEPS);
    segments.push({
      a: p,
      b: q,
      r1: radius * (1 - 0.12 * (i / STEPS)),
      r2: radius * (1 - 0.12 * ((i + 1) / STEPS)),
    });
    p = q;
  }
  if (gen === 2) {
    tips.push(p);
    return;
  }
  for (let k = 0; k < FORKS; k++) {
    // each arm forks sideways, symmetrically about its own bearing
    const phi = bearing + (k === 0 ? 1 : -1) * (Math.PI / 2.6);
    grow(p, tilt(d, SPREAD2, phi), length * 0.8, radius * TAPER, gen + 1, phi);
  }
}

// The base runs BELOW the table and gets cut off by an exact boolean later.
// Meeting the ground inside the SDF instead (max(d, -z)) makes the field graze
// the plane, and marching tetrahedra answers a graze with a rim of slivers:
// 43 edges of the first version welded into non-manifold junk, every one of
// them on this rim. Cut flat surfaces with booleans, curve them with fields.
segments.push({a: [0, 0, -0.3], b: [0, 0, 0.06], r1: BASE_R, r2: BASE_R});
segments.push({a: [0, 0, 0.06], b: [0, 0, BASE_H], r1: BASE_R, r2: BASE_R * 0.6});
segments.push({a: [0, 0, BASE_H], b: [0, 0, TRUNK_TOP],
               r1: TRUNK_R * 1.6, r2: TRUNK_R});
// Trunk top: three arms, 42 degrees off vertical, 120 degrees apart.
for (let k = 0; k < ARMS; k++) {
  const phi = (k * 2 * Math.PI) / ARMS;
  grow([0, 0, TRUNK_TOP], tilt([0, 0, 1], SPREAD1, phi), 1.25,
       TRUNK_R * TAPER, 1, phi);
}
// A knuckle above each tip. It is deliberately FATTER than the cup shank it
// meets (0.15 vs CUP_R 0.115) so the two surfaces cross transversally. A stub
// slightly thinner than the shank instead produces a near-tangential contact,
// which Manifold accepts (it keeps its own merge vectors) but which welds into
// a non-manifold edge in any downstream consumer.
for (const t of tips) {
  segments.push({a: t, b: [t[0], t[1], t[2] + 0.06], r1: 0.15, r2: 0.14});
}

// Adjacent pans must not graze each other for the same reason: two r=0.28
// pans 0.56 apart touch tangentially. Fail loudly rather than ship a pinch.
let minGap = Infinity;
for (let i = 0; i < tips.length; i++)
  for (let j = i + 1; j < tips.length; j++)
    minGap = Math.min(minGap, Math.hypot(tips[i][0] - tips[j][0],
                                         tips[i][1] - tips[j][1],
                                         tips[i][2] - tips[j][2]));
if (minGap < 2 * PAN_R + 0.12) {
  throw new Error(`tips only ${minGap.toFixed(3)} apart; pans (r ${PAN_R}) ` +
                  `would grze. Widen SPREAD2.`);
}

// ------------------------------------------------------------------- sdf ---
function sdf(p) {
  let d = Infinity;
  for (const s of segments) {
    const {d: dist, t} = segDist(p, s.a, s.b);
    const seg = dist - (s.r1 + (s.r2 - s.r1) * t);
    d = d === Infinity ? seg : smin(d, seg, 0.10);
  }
  return d;
}

/** One reusable pan + cup assembly, built once and placed six times.
 *  Deliberately over-engaged: the cup's shank runs down to the pan's bottom
 *  face and the bore floor sits well above the pan's dish, so the union is
 *  solid metal instead of two shells touching along a ring. Ring contact is
 *  what made the first attempt come out at genus 14. */
function panAndCup() {
  const pan = Manifold.cylinder(PAN_H, PAN_R, PAN_R * 0.92, 48)
                  .subtract(Manifold.sphere(0.55, 64)
                                .translate(0, 0, PAN_H + 0.55 - 0.035));
  const shank = Manifold.cylinder(PAN_H + CUP_H, CUP_R, CUP_R * 0.94, 32);
  const bore = Manifold.cylinder(0.2, BORE_R, BORE_R, 32)
                   .translate(0, 0, PAN_H + CUP_H - 0.12);
  return Manifold.union(pan, shank).subtract(bore);
}

function build() {
  const tree = Manifold.levelSet((p) => -sdf(p),
                                 {min: [-1.7, -1.7, -0.45], max: [1.7, 1.7, 3.4]},
                                 EDGE, 0, EDGE / 8);
  const table = Manifold.cube([6, 6, 2], true).translate(0, 0, -1);
  const fitting = panAndCup();
  const fittings = tips.map((t) => fitting.translate(t[0], t[1], t[2] + 0.02));
  const whole = Manifold.union([tree.subtract(table), ...fittings]);
  const lean = SIMPLIFY_TOL > 0 ? whole.simplify(SIMPLIFY_TOL) : whole;
  console.log(`  simplify: ${whole.numTri()} -> ${lean.numTri()} tris ` +
              `(tolerance ${SIMPLIFY_TOL})`);
  return lean.asOriginal().calculateNormals(3, 55);
}

const lap = stopwatch();
const candelabra = build();
describe(`B5 candelabra (edge=${EDGE}, ${segments.length} segments, ${tips.length} tips, min tip gap ${minGap.toFixed(3)})`,
         candelabra);
console.log('TIPS ' + JSON.stringify(tips.map((t) => t.map((v) => +v.toFixed(4)))));
await save(
    {
      manifold: candelabra,
      name: 'candelabra',
      material: {
        name: 'brass',
        attributes: ['NORMAL'],
        baseColorFactor: [0.72, 0.55, 0.24],
        metallic: 0.9,
        roughness: 0.35,
      },
    },
    process.argv[2]);
console.log(`bake_seconds ${lap().toFixed(2)}`);
